### PR TITLE
Issue 243: Authenticate with Backdrop in kcfinder

### PR DESCRIPTION
**Pull request by herbdool**
Addresses https://lab.civicrm.org/dev/core/issues/243

### CODE REVIEW

**Comment by xurizaemon** [comment]
At a glance this looks like the right fix for Backdrop, but we really have to keep on the radar eliminating all entry points that aren't the supported CMS authentication. Duplicating URL, path and CMS/auth resolution code through a handful of other endpoints is not sustainable.

**Comment by herbdool** [comment]
@xurizaemon sounds like a good follow-up issue. Or is there one already?

**Comment by xurizaemon** [comment]
I feel like there was one but I could never find it in JIRA. May suit https://lab.civicrm.org/dev/cloud-native/issues

**Comment by herbdool** [comment]
I think the follow-up is a general problem not just for cloud native. It needs to be standardized on the CMS classes.

**Comment by xurizaemon** [comment]
I agree it's a general problem - but I'm using cloud-native as a best practice wishlist here, rather than dropping it on the general issue list.

Totally happy for the issue to be moved or duplicated from where it is now @ https://lab.civicrm.org/dev/cloud-native/issues/16

**Comment by herbdool** [comment]
Sounds good. Let's leave it there. So I'm not sure if someone else is able to test this bug fix or not. It's pretty much identical to the Drupal function but it also checks for the regular location for modules in /modules/ and not just /sites/*/.

File: kcfinder/integration/civicrm.php [hunk 1]
```diff
@@ -108,6 +111,39 @@ function authenticate_drupal($config) {
...
+  $pos = strpos($base_url, '/sites/');
+  if ($pos === FALSE) {
+    $pos = strpos($base_url, '/profiles/');
+  }
+  if ($pos === FALSE) {
+    $pos = strpos($base_url, '/modules/');
+  }
+  $base_url = substr($base_url, 0, $pos); // Backdrop root absolute url
+
+  CRM_Utils_System::loadBootStrap(CRM_Core_DAO::$_nullArray, true, false);
```
Comment (eileenmcnaughton) [inline kcfinder/integration/civicrm.php]: looks like we can just pass an empty array without having to use the static. Note that in rare cases CRM_Core_DAO::$_nullArray can become contaminated.

**Comment by eileenmcnaughton** [comment]
So my feeling is we might struggle to get reviewers on this but 
a) it's fixing a hard error so
b) it would only be worse than existing if it introduced a security flaw - from what I can tell it is equivalent to the drupal function so would not be 'worse' and there is a permission check. I'm going to merge this 

Obviously I agree with @xurizaemon on really wanting to get rid of it - although I tried & failed not that long ago https://github.com/civicrm/civicrm-core/compare/master...eileenmcnaughton:ck?expand=1

**Comment by laryn** [comment]
I just tested with Backdrop 1.10.1 and CiviCRM 5.4.0 and it works.

**Comment by eileenmcnaughton** [comment]
Thanks @laryn

**Comment by laryn** [comment]
@eileenmcnaughton Quick followup to make sure I was clear, as I notice this was merged some time back. The patch is not included in the latest Backdrop download and I had to apply the patch manually. Is that expected?

**Comment by eileenmcnaughton** [comment]
@laryn looks like I merged on Jul 8 - that is after 1st Wed of July so it would have gone into the rc cut on first Wed of Aug - so try the 5.5 rc from https://download.civicrm.org/latest

**Comment by laryn** [comment]
Ah, yes. Thanks!!

**Pull request merged**